akka-docs-dev/rst/java/stream-cookbook.rst

```diff
@@ -107,23 +107,30 @@ Implementing reduce-by-key
 elements.
 
 The "hello world" of reduce-by-key style operations is *wordcount* which we demonstrate below. Given a stream of words
-we first create a new stream ``wordStreams`` that groups the words according to the ``i -> i`` function, i.e. now
+we first create a new stream that groups the words according to the ``i -> i`` function, i.e. now
 we have a stream of streams, where every substream will serve identical words.
 
-To count the words, we need to process the stream of streams (the actual groups containing identical words). By mapping
-over the groups and using ``fold`` (remember that ``fold`` automatically materializes and runs the stream it is used
-on) we get a stream with elements of ``Future[String,Int]``. Now all we need is to flatten this stream, which
-can be achieved by calling ``mapAsync`` with ``i -> i`` identity function.
-
-There is one tricky issue to be noted here. The careful reader probably noticed that we put a ``buffer`` between the
-``mapAsync()`` operation that flattens the stream of futures and the actual stream of futures. The reason for this is
-that the substreams produced by ``groupBy()`` can only complete when the original upstream source completes. This means
-that ``mapAsync()`` cannot pull for more substreams because it still waits on folding futures to finish, but these
-futures never finish if the additional group streams are not consumed. This typical deadlock situation is resolved by
-this buffer which either able to contain all the group streams (which ensures that they are already running and folding)
-or fails with an explicit failure instead of a silent deadlock.
-
-.. includecode:: ../../../akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/cookbook/RecipeReduceByKey.java#word-count
+To count the words, we need to process the stream of streams (the actual groups
+containing identical words). ``groupBy`` returns a :class:`SubSource`, which
+means that we transform the resulting substreams directly. In this case we use
+the ``fold`` combinator to aggregate the word itself and the number of its
+occurrences within a :class:`Pair<String, Integer>`. Each substream will then
+emit one final value—precisely such a pair—when the overall input completes. As
+a last step we merge back these values from the substreams into one single
+output stream.
+
+One noteworthy detail pertains to the ``MAXIMUM_DISTINCT_WORDS`` parameter:
+this defines the breadth of the merge operation. Akka Streams is focused on
+bounded resource consumption and the number of concurrently open inputs to the
+merge operator describes the amount of resources needed by the merge itself.
+Therefore only a finite number of substreams can be active at any given time.
+If the ``groupBy`` operator emits more than this number then the stream cannot
+continue: in order to write the new word to its new substream there needs to be
+demand on that substream and ``mergeBack`` will not create this demand because
```

Comment:
Strongly against `mergeBack` as a name, especially if the method does not exist ;)